src/stores/LayoutStore/LayoutStore.ts

```diff
@@ -186,30 +196,81 @@ export class LayoutStore {
+
+        if (!this.layouts || !newName || !this.dockedLayout) {
+            appStore.alertStore.showAlert("Save layout failed! Empty layouts or name.");
+            return;
+        }
+
+        if (PresetLayout.isPreset(newName)) {
+            appStore.alertStore.showAlert("Layout name cannot be the same as system presets.");
             return;
         }
 
-        appStore.apiService.clearLayout(layoutName).then(
-            success => {
+        if (this.layoutExists(newName)) {
+            appStore.alertStore.showAlert("Layout name already exists.");
+            return;
+        }
+
+        if (!oldName || !this.layoutExists(oldName)) {
+            appStore.alertStore.showAlert(`Cannot rename layout ${oldName}! It does not exist.`);
+            return;
+        }
+
+        appStore.dialogStore.hideSaveLayoutDialog();
+
+        // save layout to layouts[] & server/local storage
+        const configToSave = this.layouts[oldName];
+        this.layouts[newName] = configToSave;
+        if (!PresetLayout.isPreset(this.layoutNameToBeSaved)) {
+            try {
+                const success = yield appStore.apiService.setLayout(newName, configToSave);
                 if (success) {
-                    delete this.layouts[layoutName];
-                    if (layoutName === this.currentLayoutName) {
-                        this.currentLayoutName = "";
+                    const success = yield appStore.apiService.clearLayout(oldName);
+                    if (success) {
+                        delete this.layouts[oldName];
                     }
+                    this.handleRenameResult(oldName, newName, success);
                 }
-                this.handleDeleteResult(layoutName, success);
-            },
-            err => {
+            } catch (err) {
                 console.log(err);
-                this.handleDeleteResult(layoutName, false);
+                this.handleRenameResult(oldName, newName, false);
             }
-        );
+        }
+    }
+
+    private handleRenameResult = (oldName: string, newName: string, success: boolean) => {
+        if (success) {
+            AppToaster.show(SuccessToast("layout-grid", `Layout ${oldName} renamed to ${newName} successfully.`, LayoutStore.ToasterTimeout));
+            if (oldName === this.currentLayoutName) {
+                this.currentLayoutName = newName;
+            }
+        } else {
+            AlertStore.Instance.showAlert("Renaming user-defined layout failed!");
+        }
     };
 
+    @flow.bound *deleteLayout(layoutName: string) {
+        const appStore = AppStore.Instance;
+        if (!layoutName || !this.layoutExists(layoutName)) {
+            appStore.alertStore.showAlert(`Cannot delete layout ${layoutName}! It does not exist.`);
+            return;
+        }
+
+        try {
+            const success = yield appStore.apiService.clearLayout(layoutName);
+            if (success) {
+                delete this.layouts[layoutName];
```

Comment:
The new changes look good! Just one minor question: do we need to preserve the lines `if (layoutName === this.currentLayoutName) {  this.currentLayoutName = ""; }` here?

Reply:
I am thinking that these lines still need to be preserved. If a user delete a layout (eg. _layout_A_) without clear `this.currentLayoutName`, and then rename another layout(eg. _layout_B_) or add a new layout with the deleted name (_layout_A_), the current layout hint will be misleading.

Reply:
Oh, I see we also have them in `handleDeleteResult`. I missed that.